app.go

```diff
@@ -160,7 +200,38 @@ func (l nopLogger) Printf(string, ...interface{}) {
 	return
 }
 
-// An App is a modular application built around dependency injection.
+// An App is a modular application built around dependency injection. Most
+// users will only need to use the New constructor and the all-in-one Run
+// convenience method. In more unusual cases, users may need to use the Err,
+// Start, Done, and Stop methods by hand instead of relying on Run.
+//
+// New creates and initializes an App. All applications begin with a
+// constructor for the Lifecycle type already registered.
+//
+// In addition to that built-in functionality, users typically pass New a
+// handful of Provide options and one or more Invoke options. The Provide
+// options teach the application how to instantiate a variety of types, and
+// the Invoke options describe how to initialize the application.
+//
+// When created, the application immediately executes all the functions passed
+// via Invoke options. To supply these functions with the parameters they
+// need, the application looks for constructors that return the appropriate
+// types; if constructors for any required types are missing or any
+// invocations return an error, the application will fail to start (and Err
+// will return a descriptive error message).
+//
+// Once all the invocations (and any required constructors) have been called,
+// New returns and the application is ready to be started using Run or Start.
+// On startup, it executes any OnStart hooks registered with its Lifecycle.
+// OnStart hooks are executed one at a time, in order, and must all complete
```

Comment:
it might be overkill, but we could reference what we mean by "in order" by referring to the docs for the `Start` method. I'll leave this up to you.

Reply:
Sounds good, added a sentence pointing users there.